internetarchive/openlibrary
# 4271/feature/add goodreads import unit test

Closes #4271 


feature

### Technical

This refactors the fetch_goodreads.POST() function by moving most of it into a separate helper function and adding a unit test for that helper function.  It also changes the type of the csv data returned by the ```web.input()``` call from a cgi.FieldStorage object to a string (by removing the ```csv``` parameter from the ```web.input()``` call), as a string was easier to mock.  This required changing references from ```i.csv.value``` to ```i.csv```.

### Testing

1. Verify you can still upload a goodreads csv after the refactoring (the new one in test_data/ should work, which is a goodreads csv with a manually added third line to test books with no ISBNs)
2. Verify the test fails with the earlier code (see the commented out lines and the uncommented ```csv_file``` line in the first code block in the Screenshot section for how to manually revert #4259 -- you can't just revert the original change because of the type change mentioned above).  Please verify that you think that manual revert is accurate.
3. Confirm the error returned by pytest in step 2 is the same error seen in https://github.com/internetarchive/openlibrary/issues/4216#issuecomment-744030834
3. Verify the test passes from HEAD (see the second code block in the Screenshot section)
4. Make a change to the test csv file and verify the test fails.  E.g.:

```
openlibrary/openlibrary/plugins/upstream$ sed -i 's/Pippi/Steve/' tests/test_data/goodreads_library_export.csv && pytest tests/test_account.py::test_process_goodreads_csv -W ignore::DeprecationWarning -vv | egrep "^E\s*(\+|-)|passed|failed"
E             -                 'Title': 'Pippi Longstocking (Pippi Långstrump, #1)',
E             +                 'Title': 'Steve Longstocking (Pippi Långstrump, #1)',
============================== 1 failed in 0.15s ===============================
```

### Screenshot


Verifying the test fails pre-#4259 (but with s/i.csv.value/i.csv/, per above):
```
openlibrary/openlibrary/plugins/upstream$ grep -A 4 "def process_goodreads_csv" account.py && pytest tests/test_account.py::test_process_goodreads_csv -W ignore::DeprecationWarning --tb line -q
def process_goodreads_csv(i):
    import csv
    #csv_payload = i.csv if isinstance(i.csv, str) else i.csv.decode()
    #csv_file = csv.reader(csv_payload.splitlines(), delimiter=',', quotechar='"')
    csv_file = csv.reader((i.csv).splitlines(), delimiter=',', quotechar='"')
F                                                                                                                                [100%]
=============================================================== FAILURES ===============================================================
/home/a/code/openlibrary/openlibrary/plugins/upstream/account.py:954: _csv.Error: iterator should return strings, not bytes (did you open the file in text mode?)
======================================================= short test summary info ========================================================
FAILED tests/test_account.py::test_process_goodreads_csv - _csv.Error: iterator should return strings, not bytes (did you open the fi...
1 failed in 0.17s
```

Verifying the test passes post-#4259:
```
openlibrary/openlibrary/plugins/upstream$ grep -A 4 "def process_goodreads_csv" account.py && pytest tests/test_account.py::test_process_goodreads_csv -W ignore::DeprecationWarning --tb line -q
def process_goodreads_csv(i):
    import csv
    csv_payload = i.csv if isinstance(i.csv, str) else i.csv.decode()
    csv_file = csv.reader(csv_payload.splitlines(), delimiter=',', quotechar='"')
    header = next(csv_file)
.                                                                                                                                [100%]
1 passed in 0.10s
```

### Stakeholders

@cdrini @cclauss @mekarpeles
